Unable to create replication for ECR on harbor running under kubernetes

What can we help you?

I have harbor v2.2.2 running on a kubernetes cluster. I see the following error when I try to add an ECR endpoint without specifying access key/secret -

```
2021-05-28T06:08:54Z [ERROR] [/replication/adapter/awsecr/adapter.go:208]: failed to ping registry https://<id>.dkr.<region>.amazonaws.com: EC2RoleRequestError: EC2RoleRequestError: no EC2 instance role found
caused by: RequestError: send request failed
caused by: Get "http://169.254.169.254/latest/meta-data/iam/security-credentials/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

I have added following environment variables to 'core' and 'jobservice' deployments -

- AWS_ROLE_ARN - Policies/Roles set in AWS to allow read access to my ECR instance
- AWS_WEB_IDENTITY_TOKEN_FILE - I am using kubernetes OIDC discovery and projected service account token
- AWS_REGION - Though it should not be required and is extracted from the registry URL.

I do not have any AWS metadata service running (e.g. KIAM) in the cluster. All IAM policies are in place in AWS for the role I am trying to use. 

I tried an 'awscli' pod with similar environment variables and the aws CLI is able to talk to my ECR using the same role ARN.


I checked https://github.com/goharbor/harbor/issues/13687 but no mention of what changes were done to the core/jobservice pods. Also, the setup explained in the issue uses KIAM to associate roles to pods, which is not my case.

Not sure why is the awecr adaptor trying to fetch roles from metadata despite having those environment variables set ?

--- a/src/pkg/reg/adapter/awsecr/auth.go
+++ b/src/pkg/reg/adapter/awsecr/auth.go
@@ -26,7 +26,6 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/aws/credentials"
-	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
 	"github.com/aws/aws-sdk-go/aws/session"
 	awsecrapi "github.com/aws/aws-sdk-go/service/ecr"
 
@@ -102,8 +101,6 @@ func getAwsSvc(region, accessKey, accessSecret string, insecure bool, forceEndpo
 			accessKey,
 			accessSecret,
 			"")
-	} else {
-		cred = ec2rolecreds.NewCredentials(sess)
 	}
 
 	config := &aws.Config{

